### PR TITLE
[DONE] Threedidepth

### PR DESCRIPTION
**Pull request by RichardB9**
Adds a 3Di Processing provider with a algorithm for creating a waterdepth raster

### CODE REVIEW

File: processing/threedidepth_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            value = self.timestamps[index]
+        else:
+            value = 0
+        lcd_value = self.format_lcd_value(value)
+        self.lcdNumber.display(lcd_value)
+
+    def format_lcd_value(self, value: float) -> str:
+        days = int(value // 86400)
+        hours = int((value // 3600) % 24)
+        minutes = int((value // 60) % 60)
```
Comment (larsclaussen) [inline processing/threedidepth_algorithm.py]: the `divmod` function is a good way to do things like. It returns both division result and remainder in a tuple (division, remainder):
Example:
```
weeks, seconds = divmod(seconds, 7*24*60*60)
days, seconds = divmod(seconds, 24*60*60)
hours, seconds = divmod(seconds, 60*60)
minutes, seconds = divmod(seconds, 60)
```

File: processing/threedidepth_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    def new_file_event(self, file_path):
+        """New file has been selected by the user
+
+        Try to read in the timestamps from the file
+        """
+        try:
+            with h5py.File(file_path, 'r') as results:
+                timestamps = results['time'].value
+                self.set_timestamps(timestamps)
+        except Exception:
```
Comment (larsclaussen) [inline processing/threedidepth_algorithm.py]: maybe good to let the user know that something went wrong?

File: processing/threedidepth_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    """
+    Calculates waterdepths for 3Di results
+    """
+
+    # Constants used to refer to parameters and outputs. They will be
+    # used when calling the algorithm from another algorithm, or when
+    # calling from the QGIS console.
+
+    MODES = [
+        Mode(MODE_LINEAR, "Linear interpolated waterdepth"),
```
Comment (larsclaussen) [inline processing/threedidepth_algorithm.py]: waterdepth --> water depth

File: processing/threedidepth_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        Mode(MODE_CONSTANT, "Waterdepth"),
+        Mode(MODE_LINEAR_S1, "Linear interpolated waterlevel"),
```
Comment (larsclaussen) [inline processing/threedidepth_algorithm.py]: waterlevel --> water level

File: processing/threedidepth_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        Mode(MODE_LIZARD, "Lizard interpolated waterdepth"),
+        Mode(MODE_LIZARD_S1, "Lizard interpolated waterlevel"),
+    ]
+
+    GRIDADMIN_INPUT = 'GRIDADMIN_INPUT'
+    RESULTS_3DI_INPUT = 'RESULTS_3DI_INPUT'
+    DEM_INPUT = 'DEM_INPUT'
+    MODE_INPUT = 'MODE_INPUT'
+    CALCULATION_STEP_INPUT = 'CALCULATION_STEP_INPUT'
+    WATERDEPTH_OUTPUT = 'WATERDEPTH_OUTPUT'
```
Comment (larsclaussen) [inline processing/threedidepth_algorithm.py]: see above

File: processing/threedidepth_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    def name(self):
+        """Returns the algorithm name, used for identifying the algorithm"""
+        return 'threedidepth'
+
+    def displayName(self):
+        """
+        Returns the translated algorithm name, which should be used for any
+        user-visible display of the algorithm name.
+        """
+        return self.tr('Waterdepth')
```
Comment (larsclaussen) [inline processing/threedidepth_algorithm.py]: see above

**Review by larsclaussen — COMMENTED** [review]
Just some small remarks - I don't see anything strange nor any mistakes. :+1:

**Review by larsclaussen — APPROVED** [review]

File: processing/threedidepth_algorithm.py (hunk 2 above)
Comment (RichardB9) [reply processing/threedidepth_algorithm.py]: Yeah good point!

**Pull request merged**